README.md
# @kobayami/math-utils

## Installation

```sh
npm install --save @kobayami/math-utils
```

## Version and License

- Latest version: 1.1.4
- License: [MIT](https://kobayami.github.io/math-utils/LICENSE.md)
- [Changes](https://kobayami.github.io/math-utils/CHANGES.md)

## Summary

Some little math helper functions that cover the following areas:

- Division and modulo
- Lerp functions (linear interpolation)
- Clamp functions
- Root and logarithm
- Sign and integer
- Rounding

## Usage Example

```ts
import { mod, divFloor, divTrunc } from "@kobayami/math-utils/lib";

const two = mod(-13, 5);
const minusThree = divFloor(-13, 5);
const minusTwo = divTrunc(-15, 5);
```

## See Also

- [API Documentation](https://kobayami.github.io/math-utils/docs/modules.html)
- [Project Homepage](https://kobayami.github.io/math-utils)
- [Project on GitHub](https://github.com/kobayami/math-utils)
- [Issues](https://github.com/kobayami/math-utils/issues)


CHANGES.md
# Release Notes

## 1.1.4

- Fixed typo in `README.md` installation instructions

## 1.1.3

- Added `isInt32`

## 1.1.2

- Added `round`
- Improved test cases + reworked test & build infrastructure
- Added `CHANGES.md`

## 1.1.1

- Reworked `README.md`

## 1.1.0

- Added `divFloor`, `divTrunc`, `frac`, `isInteger`, `higher`, `lower`, `maskMinus0`, `signum`, `toInt32`, `log`, `root`
  and more

## 1.0.0

- Initial version


spec/index.spec.ts
import {
    clamp,
    clampUnit,
    degreesToDefaultRange,
    degreesToRadians,
    divFloor,
    divTrunc,
    frac,
    getDigits,
    higher,
    isInt32,
    isInteger,
    log,
    lower,
    maskMinus0,
    mod,
    radiansToDefaultRange,
    radiansToDegrees,
    rangeToRange,
    rangeToUnit,
    root,
    round,
    signum,
    toInt32,
    uByteToUnit,
    unitToRange,
    unitToUByte,
} from '../lib';

describe('Test Suite', () => {
    it('clamp', () => {
        for (const [value, minValue, maxValue, expected] of [
            [-1, 3, 7, 3],
            [3, 3, 7, 3],
            [5, 3, 7, 5],
            [7, 3, 7, 7],
            [11, 3, 7, 7],
        ]) {
            expect(clamp(value!, minValue!, maxValue!)).toEqual(expected!);
        }
    });
    it('clampUnit', () => {
        for (const [value, expected] of [
            [-1, 0],
            [0, 0],
            [0.2, 0.2],
            [1, 1],
            [1.1, 1],
        ]) {
            expect(clampUnit(value!)).toEqual(expected!);
        }
    });
    it('rangeToUnit, unitToRange', () => {
        for (const [value, startValue, endValue, expected] of [
            [12, 20, 30, -0.8],
            [20, 20, 30, 0],
            [27, 20, 30, 0.7],
            [30, 20, 30, 1],
            [39, 20, 30, 1.9],
        ]) {
            expect(rangeToUnit(value!, startValue!, endValue!)).toEqual(expected!);
            expect(unitToRange(expected!, startValue!, endValue!)).toEqual(value!);
        }
    });
    it('rangeToRange', () => {
        for (const [value, startValueFrom, endValueFrom, startValueTo, endValueTo, expected] of [
            [12, 20, 30, 100, 200, 20],
            [20, 20, 30, 100, 200, 100],
            [27, 20, 30, 100, 200, 170],
            [30, 20, 30, 100, 200, 200],
            [39, 20, 30, 100, 200, 290],
        ]) {
            expect(rangeToRange(value!, startValueFrom!, endValueFrom!, startValueTo!, endValueTo!)).toEqual(expected!);
        }
    });
    it('unitToUByte', () => {
        for (const [value, expected] of [
            [-1, -255],
            [-0.5, -127],
            [0, 0],
            [0.125, 32],
            [0.25, 64],
            [0.5, 128],
            [0.75, 191],
            [1, 255],
            [1.25, 319],
            [1 / 3, 85],
            [2 / 3, 170],
            [20 / 255, 20],
            [20.4 / 255, 20],
            [20.5 / 255, 21],
            [20.6 / 255, 21],
        ]) {
            expect(unitToUByte(value!)).toEqual(expected!);
        }
    });
    it('uByteToUnit', () => {
        for (const [expected, value] of [
            [-1, -255],
            [-127 / 255, -127],
            [0, 0],
            [32 / 255, 32],
            [64 / 255, 64],
            [128 / 255, 128],
            [191 / 255, 191],
            [1, 255],
            [319 / 255, 319],
            [1 / 3, 85],
            [2 / 3, 170],
            [20 / 255, 20],
            [20 / 255, 20.4],
            [20 / 255, 20.5],
            [20 / 255, 20.6],
            [21 / 255, 21],
            [21 / 255, 21.4],
            [21 / 255, 21.5],
            [21 / 255, 21.6],
        ]) {
            expect(uByteToUnit(value!)).toEqual(expected!);
        }
    });
    it('maskMinus0, signum', () => {
        for (const [value, expectedMaskMinus0, expectedSignum] of [
            [-3.4, -3.4, -1],
            [-3, -3, -1],
            [-1, -1, -1],
            [-0.2, -0.2, -1],
            [-0, 0, 0],
            [0, 0, 0],
            [3, 3, 1],
            [5.2, 5.2, 1],
        ]) {
            expect(maskMinus0(value!)).toEqual(expectedMaskMinus0!);
            expect(signum(value!)).toEqual(expectedSignum!);
        }
    });
    it('mod, divFloor, divTrunc', () => {
        for (const [a, b, expectedMod, expectedDivFloor, expectedDivTrunc] of [
            [-13, 5, 2, -3, -2],
            [-8, 5, 2, -2, -1],
            [-3, 5, 2, -1, 0],
            [2, 5, 2, 0, 0],
            [7, 5, 2, 1, 1],
            [-16, 3, 2, -6, -5],
            [-17, 3, 1, -6, -5],
            [-18, 3, 0, -6, -6],
            [-19, 3, 2, -7, -6],
            [16, 3, 1, 5, 5],
            [17, 3, 2, 5, 5],
            [18, 3, 0, 6, 6],
            [19, 3, 1, 6, 6],
        ]) {
            expect(mod(a!, b!)).toEqual(expectedMod!);
            expect(divFloor(a!, b!)).toEqual(expectedDivFloor!);
            expect(divTrunc(a!, b!)).toEqual(expectedDivTrunc!);
            expect(b! * divFloor(a!, b!) + mod(a!, b!)).toEqual(a!);
            expect(b! * divTrunc(a!, b!) + (a! % b!)).toEqual(a!);
        }
    });
    it('degreesToRadians, radiansToDegrees, degreesToDefaultRange, radiansToDefaultRange', () => {
        for (const [degrees, radians] of [
            [-0, -0 * Math.PI],
            [0, 0],
            [45, 0.25 * Math.PI],
            [90, 0.5 * Math.PI],
            [180, Math.PI],
            [270, 1.5 * Math.PI],
            [315, 1.75 * Math.PI],
        ]) {
            for (let k = -3; k < 4; k++) {
                const kDegrees = degrees! + k * 360;
                const kRadians = radians! + k * 2 * Math.PI;
                expect(degreesToRadians(kDegrees)).toBeCloseTo(kRadians, 12);
                expect(radiansToDegrees(kRadians)).toBeCloseTo(kDegrees, 12);
                expect(degreesToDefaultRange(kDegrees)).toBeCloseTo(degrees!, 12);
                expect(radiansToDefaultRange(kRadians)).toBeCloseTo(radians!, 12);
            }
        }
    });
    it('log, root', () => {
        for (const [radix, exponent, value] of [
            [2, 3, 8],
            [2, 6, 64],
            [2, 16, 0x10000],
            [3, 0, 1],
            [3, 1, 3],
            [3, 2, 9],
            [3, 3, 27],
            [3, 4, 81],
            [3, 5, 243],
            [10, -2, 0.01],
            [10, -1, 0.1],
            [10, 0, 1],
            [10, 1, 10],
            [10, 2, 100],
            [10, 3, 1000],
            [10, 9, 1000000000],
            [Math.SQRT2, 2, 2],
            [Math.SQRT2, 10, 32],
        ]) {
            expect(log(value!, radix!)).toBeCloseTo(exponent!);
            if (exponent !== 0) {
                expect(root(value!, exponent!)).toBeCloseTo(radix!);
            }
        }
    });
    it('log, root', () => {
        for (const [radix, value, expected] of [
            [10, 0, 1],
            [10, 0.002342, 1],
            [10, 0.5, 1],
            [10, 0.9992, 1],
            [10, 1, 1],
            [10, 1.000001, 1],
            [10, 1000001, 7],
            [10, 5, 1],
            [10, 9.999, 1],
            [10, 10.005, 2],
            [10, 32, 2],
            [10, 99, 2],
            [10, 99.999, 2],
            [10, 10, 2],
            [10, 100, 3],
            [10, 1000000000, 9],
            [16, 0xffff, 4],
            [16, 0xfff, 3],
            [16, 0xff, 2],
            [16, 0xf, 1],
            [16, 1, 1],
            [16, 0, 1],
            [16, 0x10000, 5],
            [16, 0x1000, 4],
            [16, 0x100, 3],
            [16, 0x10, 2],
            [16, 0x4abe2, 5],
            [16, 0x9ad3, 4],
            [16, 0xcf7, 3],
            [16, 0x7f, 2],
            [3, 3 * 3 * 3 * 3, 5],
            [3, 3 * 3 * 3 * 3 + 1, 5],
            [3, 3 * 3 * 3 * 3 - 1, 4],
        ]) {
            expect(getDigits(value!, radix!)).toBe(expected!);
        }
    });
    it('frac, round, higher, lower', () => {
        for (const [
            value,
            expectedFrac,
            expectedHigher,
            expectedLower,
            expectedRound,
            expectedRound2,
            expectedRoundMinus2,
        ] of [
            [-1.33323, -0.33323, -1, -2, -1, -1.33, -0],
            [73.1355, 0.1355, 74, 73, 73, 73.14, 100],
            [-73.1355, -0.1355, -73, -74, -73, -73.14, -100],
            [0, 0, 1, -1, 0, 0, 0],
            [0.2424, 0.2424, 1, 0, 0, 0.24, 0],
            [0.8474, 0.8474, 1, 0, 1, 0.85, 0],
            [1, 0, 2, 0, 1, 1, 0],
            [132.233, 0.233, 133, 132, 132, 132.23, 100],
            [193.923, 0.923, 194, 193, 194, 193.92, 200],
            [-0.2424, -0.2424, 0, -1, 0, -0.24, 0],
            [-0.8434, -0.8434, 0, -1, -1, -0.84, 0],
            [-1, 0, 0, -2, -1, -1, 0],
            [-132.233, -0.233, -132, -133, -132, -132.233, -100],
            [-193.923, -0.923, -193, -194, -194, -193.92, -200],
            [-1000, 0, -999, -1001, -1000, -1000, -1000],
        ]) {
            expect(frac(value!)).toBeCloseTo(expectedFrac!);
            expect(higher(value!)).toBeCloseTo(expectedHigher!);
            expect(lower(value!)).toBeCloseTo(expectedLower!);
            expect(round(value!)).toBeCloseTo(expectedRound!);
            expect(round(value!, 2)).toBeCloseTo(expectedRound2!);
            expect(round(value!, -2)).toBeCloseTo(expectedRoundMinus2!);
        }
    });
    it('round', () => {
        for (const [value, expectedRoundMinus2, expectedRoundMinus4] of [
            [0xcafebabe, 0xcafebb00, 0xcaff0000],
            [0x477f, 0x4700, 0],
            [0x477ff, 0x47800, 0x40000],
            [0x46000, 0x46000, 0x40000],
            [0x47a2c, 0x47a00, 0x40000],
            [0x48721, 0x48700, 0x50000],
            [-1, 0, 0],
        ]) {
            expect(round(value!, -2, 16)).toBeCloseTo(expectedRoundMinus2!);
            expect(round(value!, -4, 16)).toBeCloseTo(expectedRoundMinus4!);
        }
    });
    it('isInteger, toInt32, isInt32', () => {
        for (const [value, expectedToInt32, expectedIsInteger] of [
            [0, 0, 1],
            [0.827, 0, 0],
            [1, 1, 1],
            [7.34, 7, 0],
            [10, 10, 1],
            [99.9999, 99, 0],
            [100, 100, 1],
            [100.0001, 100, 0],
            [-932.9999, -932, 0],
            [-932.0001, -932, 0],
            [-932.5001, -932, 0],
            [-932.4999, -932, 0],
            [-932.0, -932, 1],
            [0x7fffffff, 0x7fffffff, 1],
            [0x80000000, -0x80000000, 1],
            [-0x80000000, -0x80000000, 1],
            [-0x80000001, 0x7fffffff, 1],
        ]) {
            expect(toInt32(value!)).toBe(expectedToInt32!);
            expect(isInt32(value!)).toBe(value === expectedToInt32!);
            expect(isInteger(value!)).toBe(!!expectedIsInteger);
        }
    });
});


src/index.ts
/**
 * Constrains a value to lie between a minimum and a maximum value.
 *
 * @param value value to constrain
 * @param minValue minimum value (inclusive)
 * @param maxValue maximum value (inclusive)
 */
export function clamp(value: number, minValue: number, maxValue: number): number {
    return Math.max(minValue, Math.min(value, maxValue));
}

/**
 * Constrains a value to lie between _0_ and _1_ (inclusive).
 *
 * @param value value to constrain
 */
export function clampUnit(value: number): number {
    return clamp(value, 0, 1);
}

/**
 * Converts a value from range _0 .. 1_ (inclusive)
 * to range _startValueTo .. endValueTo_ (inclusive)
 * by linear interpolation.
 *
 * This is equivalent to the [Lerp](https://en.wikipedia.org/wiki/Linear_interpolation) function,
 * where _value_ is the interpolation ratio, and the return value is the interpolated value.
 *
 * @param value value to convert, or interpolation ratio
 * @param startValueTo start value of the target range
 * @param endValueTo end value of the target range
 */
export function unitToRange(value: number, startValueTo: number, endValueTo: number): number {
    return startValueTo + value * (endValueTo - startValueTo);
}

/**
 * Converts a value from range _startValueFrom .. endValueFrom_ (inclusive)
 * to range _0 .. 1_ (inclusive)
 * by linear interpolation.
 *
 * This is equivalent to the inverse of the [Lerp](https://en.wikipedia.org/wiki/Linear_interpolation) function,
 * where _value_ is the interpolated value, and the return value is the interpolation ratio.
 *
 * @param value value to convert, or interpolated value
 * @param startValueFrom start value of the source range
 * @param endValueFrom end value of the source range
 */
export function rangeToUnit(value: number, startValueFrom: number, endValueFrom: number): number {
    return (value - startValueFrom) / (endValueFrom - startValueFrom);
}

/**
 * Converts a value from range _startValueFrom .. endValueFrom_ (inclusive)
 * to range _startValueTo .. endValueTo_ (inclusive)
 * by linear interpolation.
 *
 * @param value value to convert, or interpolated value
 * @param startValueFrom start value of the source range
 * @param endValueFrom end value of the source range
 * @param startValueTo start value of the target range
 * @param endValueTo end value of the target range
 */
export function rangeToRange(
    value: number,
    startValueFrom: number,
    endValueFrom: number,
    startValueTo: number,
    endValueTo: number,
): number {
    return unitToRange(rangeToUnit(value, startValueFrom, endValueFrom), startValueTo, endValueTo);
}

/**
 * Converts a value from range _0 .. 1_ (inclusive)
 * to an integer number in the range _0 .. 255_ (inclusive)
 * by linear interpolation.
 *
 * @param value value to convert, or interpolation ratio
 */
export function unitToUByte(value: number): number {
    return Math.round(unitToRange(value, 0, 255));
}

/**
 * Converts an integer number in the range _0 .. 255_ (inclusive)
 * to range _0 .. 1_ (inclusive)
 * by linear interpolation.
 *
 * @param value integer number to convert, or interpolated value
 */
export function uByteToUnit(value: number): number {
    return rangeToUnit(toInt32(value), 0, 255);
}

/**
 * Modulo operation with an always positive remainder.
 * In contrast, in _a % b_ the remainder has the sign of the dividend.
 *
 * @example
 *
 * ```ts
 * -13 % 5 === -3
 * // but:
 * mod(-13, 5) === 2
 * ```
 *
 * @param a dividend
 * @param b divisor
 */
export function mod(a: number, b: number): number {
    const r = maskMinus0(a % b);
    return r < 0 ? r + b : r;
}

/**
 * Divide operation that returns only the integer part of the quotient,
 * such that _a = div(a, b) * b + mod(a, b)_.
 * That is, the division result is rounded down.
 *
 * @example
 *
 * ```ts
 * divFloor(-13, 5) // equals -3
 * divFloor(-10, 5) // equals -2
 * divFloor(7, 5)   // equals 1
 * divFloor(15, 5)  // equals 3
 * ```
 *
 * @param a dividend
 * @param b divisor
 */
export function divFloor(a: number, b: number): number {
    return maskMinus0(Math.floor(a / b));
}

/**
 * Divide operation that returns only the integer part of the quotient,
 * such that _a = div(a, b) * b + a % b_.
 * That is, the division result is rounded towards zero.
 *
 * @example
 *
 * ```ts
 * divTrunc(-13, 5) // equals -2
 * divTrunc(-10, 5) // equals -2
 * divTrunc(7, 5)   // equals 1
 * divTrunc(15, 5)  // equals 3
 * ```
 *
 * @param a dividend
 * @param b divisor
 */
export function divTrunc(a: number, b: number): number {
    return maskMinus0(Math.trunc(a / b));
}

/**
 * Returns the fractional part, i.e., the decimal places of a value.
 *
 * @param value the value
 */
export function frac(value: number): number {
    return value - Math.trunc(value);
}

/**
 * Rounds a value to a number of decimal places.
 * If that number is negative, then the rightmost digits left from the decimal point are also rounded.
 *
 * @example
 *
 * ```ts
 * roundPositions(1234.5678)      // equals 1234
 * roundPositions(1234.5678, 2)   // equals 1234.57
 * roundPositions(1234.5678, -2)  // equals 1200
 * roundPositions(0x7ac3, -2, 16) // equals 0x7b00
 * ```
 *
 * @param value value to round
 * @param radix radix of the number system
 * @param digits accuracy in digits of the given radix
 */
export function round(value: number, digits: number = 0, radix: number = 10): number {
    const exponent = digits * Math.log(radix);
    return Math.round(value * Math.exp(exponent)) * Math.exp(-exponent);
}

/**
 * Returns the smallest integer that is strictly greater than the given value.
 *
 * @param value the value
 */
export function higher(value: number) {
    return isInteger(value) ? value + 1 : Math.ceil(value);
}

/**
 * Returns the greatest integer that is strictly less than the given value.
 *
 * @param value the value
 */
export function lower(value: number) {
    return isInteger(value) ? value - 1 : Math.floor(value);
}

/**
 * Returns the _exponent_, such that _value = radix^exponent_.
 *
 * @param value _radix^exponent_
 * @param radix _radix_
 */
export function log(value: number, radix: number): number {
    return Math.log(value) / Math.log(radix);
}

/**
 * Returns the _radix_, such that _value = radix^exponent_.
 *
 * @param value _radix^exponent_
 * @param exponent _exponent_
 */
export function root(value: number, exponent: number): number {
    return Math.exp(Math.log(value) / exponent);
}

/**
 * Returns the number of digits required to represent the integral part of a value
 * in a number system with the given radix.
 *
 * @param value the value
 * @param radix radix of the number system
 */
export function getDigits(value: number, radix: number): number {
    const adjustedValue = Math.max(Math.abs(value), 1);
    return higher(log(adjustedValue, radix));
}

/**
 * Returns the number of decimals required to represent the integral part of a value.
 *
 * @param value the value
 */
export function getDecimals(value: number): number {
    return getDigits(value, 10);
}

/**
 * Converts a value into a 32 bit signed integer number.
 *
 * This behaves like cutting the off the decimal places of the number (instead of rounding),
 * then taking the lower 32 bits of the result and interpreting them as a two's complement value.
 * This is equivalent to _value ^ 0_, because in JavaScript, bitwise operations use 32 bit operands.
 *
 * @param value the value
 */
export function toInt32(value: number): number {
    return value ^ 0;
}

/**
 * Returns _true_ if a value is a 32 bit integer, i.e., lies in the range _-2^31 .. 2^31-1_
 * and its fractional part is zero.
 *
 * @param value the value
 */

export function isInt32(value: number): boolean {
    return value === (value ^ 0);
}

/**
 * Returns _true_ if a value is an integer, i.e., its fractional part is 0.
 *
 * @param value the value
 */
export function isInteger(value: number): boolean {
    return frac(value) === 0;
}

/**
 * Replaces _-0_ with _0_, and otherwise just returns _value_.
 *
 * @param value value that may be _-0_
 */
export function maskMinus0(value: number): number {
    return value === -0 ? 0 : value;
}

/**
 * Returns the sign of a value, but unlike _Math.sign()_, returns _0_ for both _-0_ and _0_.
 *
 * @param value the value
 */
export function signum(value: number): number {
    return maskMinus0(Math.sign(value));
}

/**
 * Converts a value from radians into degrees.
 *
 * @param radians value in radians.
 */
export function radiansToDegrees(radians: number): number {
    return (radians * 180) / Math.PI;
}

/**
 * Converts a value from degrees into radians.
 *
 * @param degrees value in degrees.
 */
export function degreesToRadians(degrees: number): number {
    return (degrees * Math.PI) / 180;
}

/**
 * Maps a value, given in radians, into the modulus range _0_ (inclusive) to _2 * PI_ (exclusive).
 *
 * That is:
 *
 * - _-0.5 * PI_ is mapped to _1.5 * PI_
 * - _2 * PI_ is mapped to _0_
 * - _2.5 * PI_ is mapped to _0.5 * PI_
 *
 * @param radians
 */
export function radiansToDefaultRange(radians: number): number {
    return mod(radians, 2 * Math.PI);
}

/**
 * Maps a value, given in degrees, into the modulus range _0_ (inclusive) to _360_ (exclusive).
 *
 * That is:
 *
 * - _-90_ is mapped to _270_
 * - _360_ is mapped to _0_
 * - _450_ is mapped to _90_
 *
 * @param degrees
 */
export function degreesToDefaultRange(degrees: number): number {
    return mod(degrees, 360);
}
